samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -156,53 +159,68 @@ class KafkaCheckpointManager(checkpointSpec: KafkaStreamSpec,
     }
 
     val envelope = new OutgoingMessageEnvelope(checkpointSsp, keyBytes, msgBytes)
-    val retryBackoff: ExponentialSleepStrategy = new ExponentialSleepStrategy
 
-    val startTime = System.currentTimeMillis()
-    retryBackoff.run(
-      loop => {
-        systemProducer.send(taskName.getTaskName, envelope)
-        systemProducer.flush(taskName.getTaskName) // make sure it is written
+    val startTimeInMillis: Long = System.currentTimeMillis()
+    var sleepTimeInMillis: Long = 2
+    var producerException: Exception = null
+    while ((System.currentTimeMillis() - startTimeInMillis) <= MaxRetryDurationMs) {
+      val currentProducer = producerRef.get()
+      try {
+        currentProducer.send(taskName.getTaskName, envelope)
+        currentProducer.flush(taskName.getTaskName) // make sure it is written
         debug(s"Wrote checkpoint: $checkpoint for task: $taskName")
-        loop.done
-      },
-
-      (exception, loop) => {
-        if ((System.currentTimeMillis() - startTime) >= MaxRetryDurationMs) {
-          error(s"Exhausted $MaxRetryDurationMs milliseconds when writing checkpoint: $checkpoint for task: $taskName.")
-          throw new SamzaException(s"Exception when writing checkpoint: $checkpoint for task: $taskName.", exception)
-        } else {
+        return
+      } catch {
+        case exception: Exception => {
+          producerException = exception
           warn(s"Retrying failed checkpoint write to key: $key, checkpoint: $checkpoint for task: $taskName", exception)
+          val newProducer: SystemProducer = getSystemProducer()
+          if (producerRef.compareAndSet(currentProducer, newProducer)) {
+            info(s"Stopping the checkpoint SystemProducer")
+            currentProducer.stop()
+            info(s"Recreating the checkpoint SystemProducer")
+            for (taskName <- taskNames) {
+              info(s"Registering the taskName: $taskName with SystemProducer")
+              newProducer.register(taskName.getTaskName)
+            }
+            newProducer.start()
+          } else {
+              info("Producer instance was recreated by other thread. Retrying with it.")
+            }
```

Comment:
> The indentation of '{' and '}' around else is not aligned. Could this be fixed?

Samza checkstyle validation expects the alignment to this way. If I change it, then it will result in checkstyle errors. 

> In the else statement, should we wait for the producer to be initialized

Yes. There is already a wait-time for exponential backoff retry outside of the try-catch block. I think that should be sufficient to cover this case.  

>  Should we do newProducer.close() in the else statement?

Since we do not invoke `newProducer.start()` unless the atomic swap succeeds,  we do not have to call `newProducer.close()` in the else statement.

Reply:
Regarding "Yes. There is already a wait-time for exponential backoff retry outside of the try-catch block. I think that should be sufficient to cover this case", the backoff time starts at 100 ms. Is there a way that make sure that the next send() will definitely 100% be called after this producer is initialized? Otherwise, it seems bad to rely on 100 ms to resolve any race condition. It seems that any GC pause can add more than 100 ms random delay to the execution of any given thread.

Reply:
Regarding `Since we do not invoke newProducer.start() unless the atomic swap succeeds, we do not have to call newProducer.close() in the else statement`. KafkaProducer from Apache Kafka does not have `start()` API. If `producer.close()` is not called after the KafkaProducer is instantiated, specifically an IO thread inside the KafkaProducer, will not be cleared. So the current patch has the risk of opening many unnecessary producer thread over time due to send() failure.

Reply:
Regarding the `Samza checkstyle validation expects the alignment to this way. If I change it, then it will result in checkstyle errors`, it seems that we should update the checkstyle specification rather than conforming to the wrong format. If we prefer no to do fix the checkstyle with this PR, we probably should still have an open JIRA to track this issue.

Also, can you tell me how to reproduce this checkstyle error locally. I checked out a clean copy of samza and run `./gradlew checkstyleMain checkstyleTest` following the instruction in the README. It shows the following error:
```
Unable to create a Checker: cannot initialize module SuppressionFilter - Cannot set property 'file' in module SuppressionFilter to 'checkstyle/checkstyle-suppressions.xml': unable to read checkstyle/checkstyle-suppressions.xml
```

So I guess a different set of commands is used in travis-ci to do the style check for PR.

Reply:
1. Updated the indentation. Fixed the spacing of the braces. 
2. Invoked  produer.close() if we cannot atomically swap with the existing producer. 
3. Made sure that when other thread get the producer, deterministically we can claim that it is completely initialized.